pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -675,6 +675,10 @@ public void run() {
         // Take upsert snapshot before starting consuming events
         if (_partitionUpsertMetadataManager != null) {
           _partitionUpsertMetadataManager.takeSnapshot();
+          // If upsertTTL is enabled, we will remove expired primary keys from upsertMetadata after taking snapshot.
+          if (_tableConfig.getUpsertConfig().getMetadataTTL() > 0) {
+            _partitionUpsertMetadataManager.removeExpiredPrimaryKeys();
```

Comment:
Similar to `takeSnapshot()`, we can perform the check inside the metadata manager